adapters/impactify/impactify.go

```diff
@@ -0,0 +1,186 @@
+
+func (a *adapter) MakeRequests(bidRequest *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var errs []error
+	var validImps []openrtb2.Imp
+
+	var adapterRequests []*adapters.RequestData
+
+	for i := 0; i < len(bidRequest.Imp); i++ {
+		// Check if imp comes with bid floor amount defined in a foreign currency
+		if bidRequest.Imp[i].BidFloor > 0 && bidRequest.Imp[i].BidFloorCur != "" && strings.ToUpper(bidRequest.Imp[i].BidFloorCur) != "USD" {
+			// Convert to US dollars
+			convertedValue, err := reqInfo.ConvertCurrency(bidRequest.Imp[i].BidFloor, bidRequest.Imp[i].BidFloorCur, "USD")
+			if err != nil {
+				return nil, []error{err}
+			}
+			bidRequest.Imp[i].BidFloorCur = "USD"
+			bidRequest.Imp[i].BidFloor = convertedValue
+		}
+
+		var impactifyExt ImpactifyExtBidder
+
+		var defaultExt DefaultExtBidder
+		json.Unmarshal(bidRequest.Imp[i].Ext, &defaultExt)
+		impactifyExt.Impactify = defaultExt.Bidder
+		bidRequest.Imp[i].Ext, _ = json.Marshal(impactifyExt)
+
+		validImps = append(validImps, bidRequest.Imp[i])
+	}
+
+	if len(validImps) == 0 {
+		if errs != nil {
+			return nil, errs
+		} else {
+			return nil, []error{&errortypes.BadInput{
+				Message: "No valid impressions in the bid request",
+			}}
+		}
+	}
+
+	bidRequest.Imp = validImps
+
+	reqJSON, err := json.Marshal(bidRequest)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("x-openrtb-version", "2.5")
+	if bidRequest.Device != nil {
+		if bidRequest.Device.UA != "" {
+			headers.Add("User-Agent", bidRequest.Device.UA)
+		}
+		// Add IPv4 or IPv6 if available
+		if bidRequest.Device.IP != "" {
+			headers.Add("X-Forwarded-For", bidRequest.Device.IP)
+		} else if bidRequest.Device.IPv6 != "" {
+			headers.Add("X-Forwarded-For", bidRequest.Device.IPv6)
+		}
+	}
+	if bidRequest.Site != nil {
+		headers.Add("Referer", bidRequest.Site.Page)
+	}
+
+	// set user's cookie
+	if bidRequest.User != nil && bidRequest.User.BuyerUID != "" {
+		headers.Add("Cookie", "uids="+bidRequest.User.BuyerUID)
+	}
+
+	adapterRequests = append(adapterRequests, &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    reqJSON,
+		Headers: headers,
+	})
+
+	return adapterRequests, errs
+}
+
+func (a *adapter) MakeBids(
+	internalRequest *openrtb2.BidRequest,
+	externalRequest *adapters.RequestData,
+	response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	} else if response.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
+			Message: "Invalid request.",
+		}}
+	} else if response.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("unexpected HTTP status %d.", response.StatusCode),
+		}}
+	}
```

Comment:
Please add a json test for each possible condition in the if / else if ladder.

Reply:
This is currently not possible to add json tests for theses conditions because everything is doing in the adapter to avoid it. But if our bidding server have a problem (downtime or errors ...), these conditions will catch it.

Reply:
The json test files include a simulated response from the server. Please add test coverage for each of these cases by setting httpCalls.mockResponse object with the status code being tested. Please look to other adapters for examples.

Reply:
I see, it is added now.